src/mpi/coll/iallreduce/iallreduce.c

```diff
@@ -479,13 +479,24 @@ int MPIR_Iallreduce(const void *sendbuf, void *recvbuf, int count,
         mpi_errno = MPIR_Iallreduce_impl(sendbuf, recvbuf, count, datatype, op, comm_ptr, request);
     }
 
-    /* Copy out data from host recv buffer to GPU buffer */
-    if (host_recvbuf) {
-        recvbuf = in_recvbuf;
-        MPIR_Localcopy(host_recvbuf, count, datatype, recvbuf, count, datatype);
+    if (host_recvbuf == NULL)
+        /* no temporary receive buffer */
+        return mpi_errno;
+
+    if (*request == NULL || MPIR_Request_is_complete(*request)) {
+        /* operation is complete, copy the data */
+        MPIR_Localcopy(host_recvbuf, count, datatype, in_recvbuf, count, datatype);
+        MPIR_Coll_host_buffer_free(host_sendbuf, host_recvbuf);
+        return mpi_errno;
```

Comment:
In case `*request == NULL`, there is no place to store the information. We *could* allocate a complete request here. Right now that is handled one level up.

Reply:
Thinking some more, the complete request allocation optimization still requires the data to be copied here, because the struct might be shared by multiple collectives.

Reply:
> In case `*request == NULL`, there is no place to store the information. We _could_ allocate a complete request here. Right now that is handled one level up.

I see, I think that is enough reason to have a check and option to do the completion here. I think the code is fine. The code logic is essentially:
```
if no need for completion process
    we are done
elif completed already
    do the completion process here
else
    let request completion handle it
```
I think the current code this understanding is not apparent, so if you can enhance it with either comments or structure, that'd be nice.